python/ql/src/Security/CWE-327/TlsLibraryModel.qll

```diff
@@ -0,0 +1,134 @@
+import python
+import semmle.python.ApiGraphs
+import Ssl
+import PyOpenSSL
+
+/**
+ * A specific protocol version.
+ * We use this to identify a protocol.
+ */
+class ProtocolVersion extends string {
+  ProtocolVersion() { this in ["SSLv2", "SSLv3", "TLSv1", "TLSv1_1", "TLSv1_2", "TLSv1_3"] }
+
+  predicate lessThan(ProtocolVersion version) {
+    this = "SSLv2" and version = "SSLv3"
+    or
+    this = "TLSv1" and version = ["TLSv1_1", "TLSv1_2", "TLSv1_3"]
+    or
+    this = ["TLSv1", "TLSv1_1"] and version = ["TLSv1_2", "TLSv1_3"]
+    or
+    this = ["TLSv1", "TLSv1_1", "TLSv1_2"] and version = "TLSv1_3"
+  }
+
+  predicate isInsecure() { this in ["SSLv2", "SSLv3", "TLSv1", "TLSv1_1"] }
+}
+
+/** An unspecific protocol version */
+class ProtocolFamily extends string {
+  ProtocolFamily() { this in ["SSLv23", "TLS"] }
+}
+
+/** The creation of a context. */
+abstract class ContextCreation extends DataFlow::CfgNode {
+  /** Gets the requested protocol if any. */
+  abstract DataFlow::CfgNode getProtocol();
+}
+
+/** The creation of a connection from a context. */
+abstract class ConnectionCreation extends DataFlow::CfgNode {
+  /** Gets the context used to create the connection. */
+  abstract DataFlow::CfgNode getContext();
+}
+
+/** A context is being restricted on which protocols it can accepts. */
+abstract class ProtocolRestriction extends DataFlow::CfgNode {
+  /** Gets the context being restricted. */
+  abstract DataFlow::CfgNode getContext();
+
+  /** Gets the protocol version being disallowed. */
+  abstract ProtocolVersion getRestriction();
+}
+
+/** A context is being relaxed on which protocols it can accepts. */
+abstract class ProtocolUnrestriction extends DataFlow::CfgNode {
+  /** Gets the context being relaxed. */
+  abstract DataFlow::CfgNode getContext();
+
+  /** Gets the protocol version being allowed. */
+  abstract ProtocolVersion getUnrestriction();
+}
+
+abstract class UnspecificContextCreation extends ContextCreation, ProtocolUnrestriction {
```

Comment:
I feel this modeling of any context creation also being a relaxation of which protocols are allowed feels a bit weird. I get that it a "clever" shortcut to allow you to only look at `ProtocolUnrestriction` in your data flow query.

I think this is just a nitpick. I'm not terribly against it, it just feels kinda weird. (also, missing QLDoc)

Reply:
I added the qldoc. I also expanded the explanation in the qldoc for the configuration. I do wonder, though, if viewing context creations as unrestrictions should be something only the configuration does, even if ti would take some rearrangement of the code.